test/data_base.cpp

```diff
@@ -1304,40 +1310,131 @@ BOOST_AUTO_TEST_CASE(data_base__invalidate__invalidate__success)
+////     data_base_accessor instance(settings);
 
-BOOST_AUTO_TEST_CASE(data_base__reorganize__too_many_headers__failure)
-{
-}
+////     static const auto bc_settings = bc::system::settings(bc::system::config::settings::mainnet);
+////     BOOST_REQUIRE(instance.create(bc_settings.genesis_block));
+
+////     const auto block1 = read_block(MAINNET_BLOCK1);
+////     store_block_transactions(instance, block1, 1);
+
+////     // setup ends
+
+////     BOOST_REQUIRE_EQUAL(instance.index(block1.transactions()[0]), error::success);
+
+////     // When address_database::index is implemented, enable this test
+////     BOOST_REQUIRE(instance.addresses().get(block1.transactions()[0].outputs()[0].address().hash()));
+//// }
+
+/// reorganize headers
 
 BOOST_AUTO_TEST_CASE(data_base__reorganize__pop_and_push__success)
 {
+    create_directory(DIRECTORY);
+    bc::database::settings settings;
+    settings.directory = DIRECTORY;
+    settings.index_addresses = false;
+    settings.flush_writes = false;
+    settings.file_growth_rate = 42;
+    settings.block_table_buckets = 42;
+    settings.transaction_table_buckets = 42;
+    settings.address_table_buckets = 42;
+
+    data_base_accessor instance(settings);
+
+    static const auto bc_settings = bc::system::settings(bc::system::config::settings::mainnet);
+    const chain::block& genesis = bc_settings.genesis_block;
+    BOOST_REQUIRE(instance.create(genesis));
+
+    const auto& block1 = read_block(MAINNET_BLOCK1);
+    BOOST_REQUIRE_EQUAL(instance.push_header(block1.header(), 1, 100), error::success);
+    store_block_transactions(instance, block1, 1);
+    BOOST_REQUIRE_EQUAL(instance.candidate(block1), error::success);
+    test_heights(instance, 1u, 0u);
+
+    auto block2 = read_block(MAINNET_BLOCK2);
+    auto block2_header = block2.header();
+    block2_header.set_previous_block_hash(genesis.hash());
+    block2.set_header(block2_header);
+    const auto& block3 = read_block(MAINNET_BLOCK3);
+    
+    const auto outgoing_headers = std::make_shared<header_const_ptr_list>();
+    const auto incoming_headers = std::make_shared<const header_const_ptr_list>(header_const_ptr_list
+    {
+        std::make_shared<const message::header>(block2.header()),
+        std::make_shared<const message::header>(block3.header())
+    });
+
+    // setup ends
+    
+    BOOST_REQUIRE_EQUAL(instance.reorganize(config::checkpoint(genesis.hash(), 0), incoming_headers, outgoing_headers), error::success);
+
+    // test conditions
+    
+    test_heights(instance, 2u, 0u);
     // verify outgoing have right headers
+    BOOST_REQUIRE_EQUAL(outgoing_headers->size(), 1);
+    BOOST_REQUIRE(outgoing_headers->front()->hash() == block1.hash());
     // verify outgoing are all NOT in candidate index
+    BOOST_REQUIRE((instance.blocks().get(outgoing_headers->front()->hash()).state() & block_state::candidate) == 0);
     // verify incoming are all in candidate index
+    for (const auto& header_ptr: *incoming_headers)
```

Comment:
Avoid conditions (branching) in test cases. In this case enumerate each expectation.